networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py

```diff
@@ -207,39 +201,76 @@ def _release_vxlan_segment(self, network, host_config, level, segment):
             select = (session.query(models.PortBindingLevel).
                       filter_by(segment_id=segment_id, level=level))
 
-            if select.count() == 0:
-                segmentation_ids = self._segmentation_ids(host_config)
-                inside = segmentation_id in segmentation_ids
-                query = (session.query(AllocationsModel).
-                         filter_by(network_id=network_id, level=level, segment_type=segment_type,
-                                   segment_id=segment_id))
-                if inside:
-                    query.update({"network_id": None, "segment_id": None})
-                else:
-                    query.delete()
-
-                # Delete the network segment
-                query = (session.query(ml2_models.NetworkSegment).
-                         filter_by(id=segment_id, network_id=network_id, network_type=segment_type,
-                                   segmentation_id=segmentation_id, segment_index=level))
-
+            if select.count() > 0:
+                return False
+
+            segmentation_ids = self._segmentation_ids(host_config)
+            inside = segmentation_id in segmentation_ids
+            query = (session.query(AllocationsModel).
+                     filter_by(network_id=network_id, level=level, segment_type=segment_type,
+                               segment_id=segment_id))
+            if inside:
+                query.update({"network_id": None, "segment_id": None})
+            else:
                 query.delete()
 
-                return True
-        return False
+            # Delete the network segment
+            query = (session.query(ml2_models.NetworkSegment).
+                     filter_by(id=segment_id, network_id=network_id, network_type=segment_type,
+                               segmentation_id=segmentation_id, segment_index=level))
+
+            query.delete()
+
+            return True
+
+    @db_api.retry_db_errors
+    def allocate_baremetal_segment(self, network, hostgroup, physical_network, level, segmentation_id):
+        """Allocate a "baremetal segment" with pre-specified id
+
+        For such a segment we don't use an ACI allocation entry. No extra release method is required,
+        as it works the same way as with _release_vxlan_segment(). If a segment already exists for a
+        given physnet/network combination we raise an exception
+        """
+        segment_type = hostgroup.get('segment_type', 'vlan')
+        segment_physnet = hostgroup.get('physical_network')
+        network_id = network['id']
+
+        session = db_api.get_writer_session()
+        segment = session.query(ml2_models.NetworkSegment).filter_by(segmentation_id=segmentation_id,
```

Comment:
I know it's implemented similar as before, but wouldn't it make sense to use with_for_update() to lock the segment table? Thus you can ensure that no segments are parallel created by different instances of neutron.

Reply:
Ah probably this will be catched by the table constraints

Reply:
For baremetal segments we don't have an extra VLAN pool so the constraints will not protect us - I think. What we would need would be constraints on the networksegments table and that would be a bigger endeavour. The other solution would be to create an extra VLAN pool for all possible ~4095 minus blacklisted vlan ids per node, which would be quite a lot.

On the other hand... we will only have one node per hostgroup and with that only one port allocated to it per network and segment id (with portgroups being used for this, of course). So we should not end up in a race-condition-scenario here.